uniffi_bindgen/src/interface/error.rs

```diff
@@ -19,69 +19,146 @@
+//! ```
+//!
+//! A declaration in the UDL like this:
+//!
+//! ```
+//! # let ci = uniffi_bindgen::interface::ComponentInterface::from_webidl(r##"
+//! # namespace example {};
+//! [Error]
+//! interface Example {
+//!   one(i16 code);
+//!   two(string reason);
+//!   three(i32 x, i32 y);
+//! };
+//! # "##)?;
+//! # Ok::<(), anyhow::Error>(())
+//! ```
+//!
+//! Will result in an [`Error`] member with variants that have fields being added to the resulting [`ComponentInterface`]:
+//!
+//! ```
+//! # let ci = uniffi_bindgen::interface::ComponentInterface::from_webidl(r##"
+//! # namespace example {};
+//! # [Error]
+//! # interface Example {
+//! #   one();
+//! #   two(string reason);
+//! #   three(i32 x, i32 y);
+//! # };
+//! # "##)?;
+//! let err = ci.get_error_definition("Example").unwrap();
+//! assert_eq!(err.name(), "Example");
+//! assert_eq!(err.variants().len(), 3);
+//! assert_eq!(err.variants()[0].name(), "one");
+//! assert_eq!(err.variants()[1].name(), "two");
+//! assert_eq!(err.variants()[2].name(), "three");
+//! assert_eq!(err.variants()[0].fields().len(), 0);
+//! assert_eq!(err.variants()[1].fields().len(), 1);
+//! assert_eq!(err.variants()[1].fields()[0].name(), "reason");
+//! assert_eq!(err.variants()[2].fields().len(), 2);
+//! assert_eq!(err.variants()[2].fields()[0].name(), "x");
+//! assert_eq!(err.variants()[2].fields()[1].name(), "y");
+//! assert_eq!(err.is_flat(), false);
 //! # Ok::<(), anyhow::Error>(())
 //! ```
 
 use anyhow::Result;
 
 use super::{APIConverter, ComponentInterface};
+use super::enum_::{Enum, Variant};
 
 /// Represents an Error that might be thrown by functions/methods in the component interface.
 ///
 /// Errors are represented in the UDL as enums with the special `[Error]` attribute, but
-/// they're handled in the FFI very differently. We represent them using the `ffi_support::ExternError`
+/// they're handled in the FFI very differently. We create them in `uniffi::call_with_result()` if
+/// the wrapped function returns an `Err` value
 /// struct and assign an integer error code to each variant.
-#[derive(Debug, Clone, Default, Hash)]
+#[derive(Debug, Clone, Hash)]
 pub struct Error {
-    pub(super) name: String,
-    pub(super) values: Vec<String>,
+    pub name: String,
+    enum_: Enum,
 }
 
 impl Error {
+    pub fn from_enum(enum_: Enum) -> Self{
+        Self {
+            name: enum_.name.clone(),
+            enum_,
+        }
+    }
+
     pub fn name(&self) -> &str {
         &self.name
     }
 
-    pub fn values(&self) -> Vec<&str> {
-        self.values.iter().map(|v| v.as_str()).collect()
+    pub fn wrapped_enum(&self) -> &Enum {
+        &self.enum_
+    }
+
+    pub fn variants(&self) -> Vec<&Variant> {
+        self.enum_.variants()
+    }
+
+    pub fn is_flat(&self) -> bool {
+        self.enum_.is_flat()
+    }
+
+    // For compatibility with the Enum interface
+    pub fn contains_object_references(&self, _: &ComponentInterface) -> bool {
+        false
```

Comment:
This one confused me since it seemed like only object reference would be an Interface, which doesn't seem possible to nest in an Error or an Enum.  What's the correct behavior here?

Reply:
I think the right thing to do either way is to delegate to `self.enum_.contains_object_references`, and either let it continue to do its thing or update it to always return `false`. Off the top of my head I don't see anything that would *prevent* declaring an enum containing an interface type, but I haven't thought about it very hard.